### DISCUSSION TITLE
Latency and Timestamps

### DISCUSSION THREAD

**Discussion by danmoser** [FreeOpcUa/opcua-asyncio#1092] [Q&A] [answered]
Hi folks,
I need to read some values from an OPC-UA server with as little latency as possible. 
(1) Is the PubSub method the most suitable for this?

Also, I wanted to better understand timestamps within asyncua. 
(2) Why asyncua servers do not update "ServerTimestamps"? 
(3)  Why Source- and Server-timestamps are empty in data subscriptions (as below)? 
(4) Is there a way to enable them?  

```
INFO:asyncua.common.subscription:Publish callback called with result: PublishResult(SubscriptionId=81, 
AvailableSequenceNumbers=[1], MoreNotifications=False, NotificationMessage_=NotificationMessage(SequenceNumber=1, 
PublishTime=datetime.datetime(2022, 10, 18, 18, 7, 48, 937182), NotificationData=[DataChangeNotification(MonitoredItems=
[MonitoredItemNotification(ClientHandle=201, Value=DataValue(Value=Variant(Value=-0.9550239171510926, VariantType=
<VariantType.Double: 11>, Dimensions=None, is_array=False), StatusCode_=StatusCode(value=0), SourceTimestamp=None, 
ServerTimestamp=None, SourcePicoseconds=None, ServerPicoseconds=None)), MonitoredItemNotification(ClientHandle=202, 
Value=DataValue(Value=Variant(Value=None, VariantType=<VariantType.Null: 0>, Dimensions=None, is_array=False), 
StatusCode_=StatusCode(value=0), SourceTimestamp=None, ServerTimestamp=None, SourcePicoseconds=None, 
ServerPicoseconds=None))], DiagnosticInfos=[])]), Results=[], DiagnosticInfos=[])

INFO:asyncua:datachange_notification Node(NodeId(Identifier=13, NamespaceIndex=2, 
NodeIdType=<NodeIdType.FourByte: 1>)) -0.9550239171510926
```

I'm using asyncua 0.98.
Thanks in advance!

**Comment by AndreasHeine** [comment]
2: did you create the datavalue on your own? because you can set it there

3/4: depends on the CreateMonitoredItems Request -> timestampsToReturn 	   EnumTimestamps ToReturn 	   An enumeration that specifies the timestamp Attributes to be transmitted for each MonitoredItem. The TimestampsToReturn enumeration is defined in 7.40.When monitoring Events, this applies only to Event fields that are of type DataValue.

**Reply by AndreasHeine** [reply]
the lowest latency would be with OPC UA PubSub via UDP (not implemented!)

**Comment by danmoser** [comment, marked as answer]
Thank you, @AndreasHeine .
(1) Understood about the UDP. Let me know if you suggest something else in the current implementation.
(2) All the opcua-asyncio repository examples do not update ServerTimestamps (date is fixed as Dec/31/2000).

~~(3/4) I'm probably asking too much, but could you provide an example? I checked `asyncua/common/subscription.py`, and the `create_monitored_items()` already has `params.TimestampsToReturn = ua.TimestampsToReturn.Both`. I don't know what to change in order to get the Source- and Server-timestamps from the subscriptions.~~

(2/3/4) I've now learned that asyncua doesn't automatically update ServerTimestamp with `write_value()` like it does with SourceTimestamp! And that the Dec/31/2020 timestamps are actually the equivalent of `None`!
I was getting neither the Source- nor the Server-Timestamps with `subscribe_data_change()` because my server was saving the values with raw `ua.DataValue()`.

Lesson learned. Thanks!!

**Reply by AndreasHeine** [reply]
whats the target? max latency you aim at?

**Reply by danmoser** [reply]
ideally about 1ms (server-client in the same network)

**Reply by AndreasHeine** [reply]
sounds impossible with python and tcp!

**Reply by danmoser** [reply]
Agreed.

**Reply by AndreasHeine** [reply]
what you need is OPC UA PubSub over TSN which has a cycletime of <2 ms means special requirements for the used network hardware...
